bind.sh

```diff
@@ -13,92 +13,110 @@
+      exit 1
+  fi
 fi
 
-# Read binding specifications
-while [[ $# -gt 0 ]]; do
-    case "$1" in
-        --cpus)
-            CPUS=(${2//\// })
-            if [[ "$IDX" -ge "${#CPUS[@]}" ]]; then
-                echo "Error: Incomplete CPU binding specification" 1>&2
-                exit 1
-            fi
-            ;;
-        --gpus)
-            GPUS=(${2//\// })
-            if [[ "$IDX" -ge "${#GPUS[@]}" ]]; then
-                echo "Error: Incomplete GPU binding specification" 1>&2
-                exit 1
-            fi
-            ;;
-        --mems)
-            MEMS=(${2//\// })
-            if [[ "$IDX" -ge "${#MEMS[@]}" ]]; then
-                echo "Error: Incomplete MEM binding specification" 1>&2
-                exit 1
-            fi
-            ;;
-        --nics)
-            NICS=(${2//\// })
-            if [[ "$IDX" -ge "${#NICS[@]}" ]]; then
-                echo "Error: Incomplete NIC binding specification" 1>&2
-                exit 1
-            fi
-            ;;
-        *)
-            break
-            ;;
-    esac
-    shift 2
-done
+if [ -n "${gpus+x}" ]; then
+  gpus=(${gpus//\// })
+  if [[ "$rank" -ge "${#gpus[@]}" ]]; then
+      echo "Error: Incomplete GPU binding specification" 1>&2
+      exit 1
+  fi
+  export CUDA_VISIBLE_DEVICES="${gpus[$rank]}"
+fi
 
-# Prepare environment
-if [[ -n "${GPUS+x}" ]]; then
-    export CUDA_VISIBLE_DEVICES="${GPUS[$IDX]}"
+if [ -n "${mems+x}" ]; then
+  mems=(${mems//\// })
+  if [[ "$rank" -ge "${#mems[@]}" ]]; then
+      echo "Error: Incomplete MEM binding specification" 1>&2
+      exit 1
+  fi
 fi
-if [[ -n "${NICS+x}" ]]; then
-    # Set all potentially relevant variables, hopefully they are ignored if we
-    # are not using the corresponding network.
-    NIC="${NICS[$IDX]}"
-    export UCX_NET_DEVICES="${NIC//,/:1,}":1
-    export NCCL_IB_HCA="$NIC"
-    NIC_ARR=(${NIC//,/ })
-    export GASNET_NUM_QPS="${#NIC_ARR[@]}"
-    export GASNET_IBV_PORTS="${NIC//,/+}"
+
+if [ -n "${nics+x}" ]; then
+  nics=(${nics//\// })
+  if [[ "$rank" -ge "${#nics[@]}" ]]; then
+      echo "Error: Incomplete NIC binding specification" 1>&2
+      exit 1
+  fi
+
+  # Set all potentially relevant variables (hopefully they are ignored if we
+  # are not using the corresponding network).
+  nic="${nics[$rank]}"
+  nic_array=(${nic//,/ })
+  export UCX_NET_DEVICES="${nic//,/:1,}":1
+  export NCCL_IB_HCA="$nic"
+  export GASNET_NUM_QPS="${#nic_array[@]}"
+  export GASNET_IBV_PORTS="${nic//,/+}"
 fi
 
-# Prepare command
-if command -v numactl &> /dev/null; then
-    if [[ -n "${CPUS+x}" ]]; then
-        set -- --physcpubind "${CPUS[$IDX]}" "$@"
-    fi
-    if [[ -n "${MEMS+x}" ]]; then
-        set -- --membind "${MEMS[$IDX]}" "$@"
-    fi
-    set -- numactl "$@"
-elif [[ -n "${CPUS+x}" || -n "${MEMS+x}" ]]; then
-    echo "Warning: numactl is not available, cannot bind to cores or memories" 1>&2
+if [ $launcher != "local" ]; then
```

Comment:
@manopapad please confirm, when launcher is "local", all of this should be skipped.

Reply:
We would like the binding to still happen, even if it's just a singleton invocation. As a matter of fact, `test.py` depends on this behavior.

Reply:
Right, I guess I was thinking of this case:
```
    if all(x is None for x in (cpu_bind, mem_bind, gpu_bind, nic_bind)):
        return ()
``` 
which suppressed `bind.sh` at all if no bind args were set. We could:

* detect this inside `bind.sh` and skip the `numactl` invocation
* detect this in the driver and pass `--launcher=none` to flag skipping `numactl`

Do you have a preference?

Reply:
`numactl` should always be used if `--cpu-bind` or `--mem-bind` is present (and `numactl` is available on the system), regardless of the value of `--launcher`. So I think we need to check for the presence of either of these options inside `bind.sh`.

Reply:
https://github.com/nv-legate/legate.core/pull/461/commits/cb05f269f6a078b470b319e448f3e558d261aa64